### PR TITLE
Factories and data generation

### LINKED ISSUE #304: Test data for development
It's very hard to work on modernomad without any test data. This could be a sanitised version of some production data, or some manually created test data.

### PR DESCRIPTION
**Pull request by jonathan-s**
Work in progress for factories that can then generate data. 

You can now generate data with the command `python manage.py generate_test_data`

Closes #304

### CODE REVIEW

**Comment by bfirsh** [comment]
I'm getting a traceback. Seems to be from the sqlflush:

```
Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 345, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 348, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 399, in execute
    output = self.handle(*args, **options)
  File "/app/core/management/commands/generate_test_data.py", line 28, in handle
    cursor.execute(statement)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/utils.py", line 95, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/usr/local/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 62, in execute
    return self.cursor.execute(sql)
django.db.utils.ProgrammingError: can't execute an empty query
```

This is the value of `flush_db`:

```sql
BEGIN;
TRUNCATE "core_location_readonly_admins", "core_baseimage", "auth_user_groups", "core_locationimage", "bank_currency", "core_userprofile", "djcelery_periodictask", "gather_event_attendees", "core_emailtemplate", "gather_eventnotifications_location_publish", "core_roomimage", "core_locationemailtemplate", "core_usenote", "django_admin_log", "core_usetransaction", "django_session", "core_bookingbill", "djcelery_periodictasks", "bank_account_admins", "gather_event_endorsements", "gather_event", "core_backing", "djcelery_intervalschedule", "core_bill", "gather_eventnotifications_location_weekly", "core_payment", "core_locationmenu", "auth_user", "core_booking_suppressed_fees", "core_billlineitem", "django_flatpage_sites", "core_resource", "core_fee", "gather_event_organizers", "core_location", "django_content_type", "core_capacitychange", "bank_systemaccounts", "bank_transaction", "core_locationfee", "auth_group", "django_site", "django_flatpage", "bank_account_owners", "celery_tasksetmeta", "auth_permission", "core_subscriptionbill", "gather_eventnotifications", "core_userprofile_primary_accounts", "gather_eventadmingroup_users", "core_use", "celery_taskmeta", "djcelery_taskstate", "core_houseaccount", "bank_account", "core_subscription", "core_usernote", "bank_entry", "auth_group_permissions", "core_subscriptionnote", "core_booking", "core_backing_users", "core_location_house_admins", "core_locationflatpage", "djcelery_workerstate", "gather_eventadmingroup", "gather_eventseries", "auth_user_user_permissions", "djcelery_crontabschedule";
SELECT setval(pg_get_serial_sequence('"core_location"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_resource"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_fee"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_bill"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_subscription"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_use"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_booking"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_payment"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_userprofile"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_emailtemplate"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_locationemailtemplate"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_locationfee"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_billlineitem"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_locationmenu"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_locationflatpage"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_usernote"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_usenote"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_subscriptionnote"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_baseimage"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_capacitychange"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_backing"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_houseaccount"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"core_usetransaction"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"bank_currency"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"bank_account"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"bank_systemaccounts"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"bank_transaction"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"bank_entry"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"celery_taskmeta"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"celery_tasksetmeta"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"djcelery_intervalschedule"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"djcelery_crontabschedule"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"djcelery_periodictask"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"djcelery_workerstate"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"djcelery_taskstate"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"gather_eventadmingroup"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"gather_eventseries"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"gather_event"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"gather_eventnotifications"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"auth_permission"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"auth_group"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"auth_user"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"django_content_type"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"django_site"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"django_admin_log"','id'), 1, false);
SELECT setval(pg_get_serial_sequence('"django_flatpage"','id'), 1, false);

COMMIT;
```

File: core/factory_apps/location.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    location = factory.SubFactory(LocationFactory)
+    fee = factory.SubFactory(FeeFactory)
+
+
+class LocationMenuFactory(factory.DjangoModelFactory):
+    class Meta:
+        model = LocationMenu
+
+    location = factory.SubFactory(LocationFactory)
+    name = factory.Faker('text')
```
Comment (bfirsh) [inline core/factory_apps/location.py]: I'm getting `value too long for type character varying(15)` errors:

```suggestion
    name = factory.Faker('text', max_nb_chars=15)
```

**Comment by bfirsh** [comment]
Two extra things:

- Would be great to get this included as part of the getting started docs
- It could probably be run as part of the tests to make sure it at least runs without error

File: core/factory_apps/location.py (hunk 1 above)
Comment (jonathan-s) [reply core/factory_apps/location.py]: Added in the last commit.

**Comment by jonathan-s** [comment]
@bfirsh Fixed the postgres issues, a test has been added for the command and some documentation has been added as well.

**Comment by bfirsh** [comment]
Awesome! Works great now. :tada:

**Comment by bfirsh** [comment]
Oh, funnn... the factory library doesn't seem to work when media is on S3, so it can't generate data for Heroku review apps. :|

I have no idea what it is trying to do, but uploading images with the admin works just fine.

```
Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 345, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 348, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 399, in execute
    output = self.handle(*args, **options)
  File "/app/core/management/commands/generate_test_data.py", line 38, in handle
    locationobj = location.LocationFactory()
  File "/usr/local/lib/python3.6/site-packages/factory/base.py", line 46, in __call__
    return cls.create(**kwargs)
  File "/usr/local/lib/python3.6/site-packages/factory/base.py", line 563, in create
    return cls._generate(enums.CREATE_STRATEGY, kwargs)
  File "/usr/local/lib/python3.6/site-packages/factory/base.py", line 500, in _generate
    return step.build()
  File "/usr/local/lib/python3.6/site-packages/factory/builder.py", line 279, in build
    kwargs=kwargs,
  File "/usr/local/lib/python3.6/site-packages/factory/base.py", line 314, in instantiate
    return self.factory._create(model, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/factory/django.py", line 163, in _create
    return cls._get_or_create(model_class, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/factory/django.py", line 154, in _get_or_create
    instance, _created = manager.get_or_create(*args, **key_fields)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/manager.py", line 122, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 467, in get_or_create
    return self._create_object_from_params(lookup, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 499, in _create_object_from_params
    obj = self.create(**params)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 401, in create
    obj.save(force_insert=True, using=self.db)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 708, in save
    force_update=force_update, update_fields=update_fields)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 736, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 820, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 859, in _do_insert
    using=using, raw=raw)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/manager.py", line 122, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 1039, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1059, in execute_sql
    for sql, params in self.as_sql():
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1019, in as_sql
    for obj in self.query.objs
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1019, in <listcomp>
    for obj in self.query.objs
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1018, in <listcomp>
    [self.prepare_value(field, self.pre_save_val(field, obj)) for field in fields]
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 968, in pre_save_val
    return field.pre_save(obj, add=True)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/fields/files.py", line 311, in pre_save
    file.save(file.name, file, save=False)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/fields/files.py", line 106, in save
    self._size = content.size
  File "/usr/local/lib/python3.6/site-packages/django/db/models/fields/files.py", line 75, in _get_size
    return self.file.size
  File "/usr/local/lib/python3.6/site-packages/django/core/files/base.py", line 59, in _get_size
    self._size = self._get_size_from_underlying_file()
  File "/usr/local/lib/python3.6/site-packages/django/core/files/base.py", line 49, in _get_size_from_underlying_file
    pos = self.file.tell()
ValueError: I/O operation on closed file.
```

Maybe as a quick hack we could add a command-line option to disable generating images... I can do this in a follow-up PR with the Heroku stuff I'm doing if you'd like.

**Comment by jonathan-s** [comment]
@bfirsh This sounds like an issue that we should report upstream. Disabling generation of images as a commandline option sounds like a good idea. I think some images are required though ... so perhaps for those we would need to make them nullable and then use the option Blank=False so that a form would still complain if it's empty.

**Comment by bfirsh** [comment]
:+1: Or just enter a fake broken file path (for now).

**Pull request merged**